src/Stratis.Bitcoin/Consensus/BestChainIndexer.cs

```diff
@@ -0,0 +1,184 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using NBitcoin;
+
+namespace Stratis.Bitcoin.Consensus
+{
+    /// <summary>
+    /// An indexer of the consensus chain.
+    /// </summary>
+    public class BestChainIndexer
+    {
+        private readonly IConsensusManager consensusManager;
+        private readonly Dictionary<int, ChainedHeader> blocksByHeight = new Dictionary<int, ChainedHeader>();
+        private readonly object lockObj = new object();
+
+        internal BestChainIndexer(IConsensusManager consensusManager)
+        {
+            this.consensusManager = consensusManager;
+        }
+
+        public ChainedHeader Tip => this.consensusManager.Tip;
+
+        internal void Initialize(ChainedHeader chainedHeader)
+        {
+            while (chainedHeader != null)
+            {
+                this.blocksByHeight.Add(chainedHeader.Height, chainedHeader);
+                chainedHeader = chainedHeader.Previous;
+            }
+        }
+
+        /// <summary>
+        /// Add a new header to the best chain.
+        /// </summary>
+        /// <remarks>The method should only be set from <see cref="IConsensusManager"/>.</remarks>
+        /// <param name="tipToAdd">The new chained header, either a new tip or a rewind of the tip.</param>
+        internal void AddTip(ChainedHeader tipToAdd)
```

Comment:
this method and the next one can be invoked using signaling and by doing that removed from the interface of the component. By doing that we will be able to avoid manually calling those methods from CM

Reply:
I considered that, but the BCI is a complementry class ot CM, otherwise we can just put all methods in CM itself, consideirng this the problem with using the signaling is that we probably want to guarantee BCI tip be set first before block is signaled to other components.